### PR TITLE
feat: support bitbucket and gitlab shortcut, support private repo through sshurl

### PR DESCRIPTION
**Pull request by huochunpeng**
Removed usage of Github api, normalised all hosts (Github/Bitbucket/Gitlab) using git/https/ssh url.
By using sshurl, private repo now works as long as user has ssh public key setup properly on host account.
Patched Github tarball url, can be removed after https://github.com/npm/hosted-git-info/pull/34.

### CODE REVIEW

**Comment by huochunpeng** [comment]
Somehow, travis-cli failed to see public repo
```
(node:2462) UnhandledPromiseRejectionWarning: Error: Command failed: git ls-remote --refs git+ssh://git@github.com/zkochan/is-negative.git master
Permission denied (publickey).
fatal: Could not read from remote repository.
```

**Comment by huochunpeng** [comment]
@zkochan could you [configure a user key](https://docs.travis-ci.com/user/private-dependencies/) on travis-ci, apparently it's even needed for ssh url to work on public repo.

ref: https://stackoverflow.com/questions/15674064/how-to-fix-a-permission-denied-publickey-error-for-a-git-submodule-update-in-t

**Comment by zkochan** [comment]
but it means that everyone will have to make this configuration, who will use git dependencies. Currently it is not required by npm/yarn

**Comment by huochunpeng** [comment]
Yes, I am surprised too travis-ci cannot even see public repo. I will investigate bit more.

**Comment by huochunpeng** [comment]
I got idea why npm/yarn don't have problem with installing git dep on travis-ci.

When npm install git dep, it tries git url and https url before ssh url.
Both git and https url works on travis-ci. Keep in mind, only ssh url works with private repo transparently.

I guess yarn did similar thing.

```js
test('github-shortcut', function (t) {
  const cloneUrls = [
    ['git://github.com/foo/private.git', 'GitHub shortcuts try git URLs first'],
    ['https://github.com/foo/private.git', 'GitHub shortcuts try HTTPS URLs second'],
    ['ssh://git@github.com/foo/private.git', 'GitHub shortcuts try SSH third']
```

@zkochan how you think we replicate the behaviour? Try git/https/ssh url in sequence?

Performance wise, git url is fastest. I guess it's due to no encryption.
```
⋊> ~/g/debug-travis on master ◦ time git ls-remote --refs git://github.com/huochunpeng/ast-matcher.git master
        1.39 real         0.00 user         0.01 sys
⋊> ~/g/debug-travis on master ◦ time git ls-remote --refs https://github.com/huochunpeng/ast-matcher.git master
        2.73 real         0.04 user         0.02 sys
⋊> ~/g/debug-travis on master ◦ time git ls-remote --refs git+ssh://git@github.com/huochunpeng/ast-matcher.git master
        6.17 real         0.04 user         0.02 sys
```

**Comment by huochunpeng** [comment]
@zkochan There is something I don't like about https path.

When installing private repo, npm pops up a dialog to ask for username/password. Apparently it is trying https url. If I cancel the login, npm still got my private repo installed, apparently it falls back to ssh url which works.

NOT sure whether yarn tried https url before ssh url. From what I can see in yarn.lock file, yarn uses ssh url, not https url. Since yarn mocked git commands in test, it doesn't have issue in ci.

Shall we only try git url and ssh url?

**Comment by zkochan** [comment]
@andreineculau you might want to check this. You worked a lot with this repo

**Comment by huochunpeng** [comment]
I made another small change.

When try HTTP HEAD on `https://domain/user/project.git`, try url without `.git`. Because with `.git` ending, both Github and gitlab respond 302 to redirect it to `https://domain/user/project`. Need to use url without `.git` ending to properly test visibility of the repo.

BitBucket treats both urls with same response.

I am adding some basic test coverage for gitlab.

**Comment by huochunpeng** [comment]
Added test coverage on gitlab.
Added one test on private repo.

File: test/index.ts [hunk 1]
```diff
@@ -154,3 +154,177 @@ test('resolveFromGit() with commit from non-github repo', async t => {
...
+    resolution: {
+      tarball: 'https://codeload.github.com/zkochan/is-negative/tar.gz/2fa0531ab04e300a24ef4fd7fb3a280eccb7ccc5',
+    },
+    resolvedVia: 'git-repository',
+  })
+  t.end()
+});
+
+// This test relies on implementation detail.
+// current implementation does not try git ls-remote --refs on pref with full commit hash, this fake repo url will pass.
```
Comment (zkochan) [inline test/index.ts]: so what makes it private then?

Comment (huochunpeng) [reply test/index.ts]: Not this one, the one after. The fake repo fails check on both git url and https url, effectively runs the same code path of private repo by using ssh url.

**Comment by zkochan** [comment]
:ship: pnpm v2.10.0

**Pull request merged**